{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "PeptideMutationFinder.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/retnuh-k/peptide-mutation-finder/blob/main/PeptideMutationFinder.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Mutation to peptide sequence pipeline\n",
        "\n",
        "This notebook automates the process of figuring out changes in protein sequences as a result of mutations found from genome sequencing. Written by Hunter King (https://github.com/retnuh-k/) under the guidance of Dr. Songtao Jia and Yimeng Fang as part of the Jia Lab (http://jia.biology.columbia.edu/).\n",
        "\n",
        "Sample files available on github: https://github.com/retnuh-k/peptide-mutation-finder \n",
        "\n",
        "\n",
        "**Required input files:** \n",
        "*   mutation_comparison_tsv_file - file describing the differences in two vcf files outputted by vcftools --gzdiff function (see http://vcftools.sourceforge.net/man_latest.html)\n",
        "*   geneome_tsv_file - a tsv file of the organism's feature coordinates in gff3 file format (can create by removing header starting with '#' from gff3 file and saving as tsv)\n",
        "*   genome_fasta_file - a fasta file with chromosome headers written as \">[chromosome name]\"\n",
        "\n",
        "**Full pipeline steps**\n",
        "1.   Obtain sequencing data of strain of interest and wild type\n",
        "2.   Download reference genome fasta file\n",
        "3.   Run freebayes variant detector (see https://github.com/freebayes/freebayes) on both strain and wild type .bam files against the reference genome to get a vcf (variant call file) for each\n",
        "4.   Filter each vcf file by quality (recommended between 10 and 30) and (optionally) depth using vcffilter or equivalent (https://github.com/vcflib/vcflib)\n",
        "5.   gzip filtered vcf files\n",
        "6.   Compare vcf files using vcftools (http://vcftools.sourceforge.net/man_latest.html), run \"vcftools --gzvcf test_strain.vcf.gz --gzdiff wild_time.vcf.gz --diff-site --out comparison_output_file\". This step ensures that only mutations not shared with the wild type are analyzed.\n",
        "7.   Rename comparison output file to have .tsv file extension\n",
        "8.   Download genome fasta file and reformat per the above specifications\n",
        "9.   Download genome coordinate gff3 file, remove header and save as .tsv file\n",
        "10.  Upload required files to notebook, specify filenames in code block below, and run notebook \n",
        "\n"
      ],
      "metadata": {
        "id": "7-knVOKpbrDK"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title (Optional) Download sample files\n",
        "\n",
        "!git clone https://github.com/retnuh-k/peptide-mutation-finder.git"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "cellView": "form",
        "id": "YMsZGwgr2qbM",
        "outputId": "50d08f5d-dfdf-48a0-a354-6ca2795d5928"
      },
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'peptide-mutation-finder'...\n",
            "remote: Enumerating objects: 38, done.\u001b[K\n",
            "remote: Counting objects: 100% (38/38), done.\u001b[K\n",
            "remote: Compressing objects: 100% (37/37), done.\u001b[K\n",
            "remote: Total 38 (delta 18), reused 0 (delta 0), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (38/38), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "import re\n",
        "try:\n",
        "  import Bio\n",
        "except:\n",
        "  !pip install Biopython\n",
        "from Bio.Seq import Seq\n",
        "from google.colab import files\n",
        "\n",
        "#@title Input filenames, then hit Runtime -> Run All\n",
        "mutation_comparison_tsv_file = \"peptide-mutation-finder/TT_v_WT_DP10QU20.tsv\" #@param {type:\"string\"}\n",
        "geneome_tsv_file = \"peptide-mutation-finder/Schizosaccharomyces_pombe_all_chromosomes.tsv\" #@param {type:\"string\"}\n",
        "genome_fasta_file = \"peptide-mutation-finder/pombe_genome.fasta\" #@param {type:\"string\"}\n",
        "job_name = \"mutation_results\" #@param {type:\"string\"}\n",
        "download_mutation_csv = True #@param {type:\"boolean\"}\n",
        "\n",
        "# Load Dataframes\n",
        "mut_df = pd.read_csv(mutation_comparison_tsv_file,sep='\\t',header=0)\n",
        "mask = mut_df['IN_FILE']=='1'\n",
        "mut_df_1 = mut_df[mask]\n",
        "genome_df = pd.read_csv(geneome_tsv_file,\n",
        "                        sep='\\t',\n",
        "                        names=[\"chromosome\",\"source\",\"seq_type\",\"start_pos\",\"end_pos\",\"score\",\"strand\",\"frame\",\"id\"])\n",
        "\n",
        "exon_m = genome_df['id'].str.contains(\"exon\")\n",
        "exon_df = genome_df[exon_m]\n",
        "for row in exon_df.index:\n",
        "  if \"RNA\" in str(exon_df.loc[row,\"id\"]):\n",
        "    exon_df.drop(row,inplace=True)\n",
        "\n",
        "mut_df_1.insert(mut_df_1.shape[1],\"gene_id\",np.empty(mut_df_1.shape[0]))\n",
        "mut_df_1.insert(mut_df_1.shape[1],\"gene_start\",np.empty(mut_df_1.shape[0]))\n",
        "mut_df_1.insert(mut_df_1.shape[1],\"gene_end\",np.empty(mut_df_1.shape[0]))\n",
        "mut_df_1.insert(mut_df_1.shape[1],\"strand\",np.empty(mut_df_1.shape[0]))\n",
        "mut_df_1.insert(mut_df_1.shape[1],\"frame\",np.empty(mut_df_1.shape[0]))\n",
        "\n",
        "chromosome_dfs = {}\n",
        "for chrom in set(exon_df['chromosome']):\n",
        "  chrom_mask = exon_df['chromosome'] == chrom\n",
        "  chromosome_dfs[chrom] = pd.DataFrame(exon_df[chrom_mask])\n",
        "\n",
        "genome_str = \"\"\n",
        "with open(genome_fasta_file, 'r') as file:\n",
        "  for line in file:\n",
        "    if '>' in line:\n",
        "      genome_str = genome_str + \"\\n\" + line\n",
        "    else:\n",
        "      genome_str = genome_str + line[:-1]\n",
        "genome_str = genome_str[1:] # remove first newline"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yUaEc7EkX5us",
        "outputId": "307c1e8e-8d8a-4853-e136-475c5532b50b",
        "cellView": "form"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  errors=errors,\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Function definitions\n",
        "\n",
        "# Store which genes the mutations are in if they are in exons\n",
        "def binary_mutation_search(mut_df, chrom_df_arr):\n",
        "  hits = 0\n",
        "  df = mut_df.copy()\n",
        "  for i in df.index:\n",
        "    chrom_df = chrom_df_arr[mut_df_1.loc[i,'CHROM']]\n",
        "    chrom_df.sort_values(by='start_pos',axis=0,inplace=True)\n",
        "    mut_pos = int(df.loc[i,'POS1'])\n",
        "\n",
        "    lower = -1\n",
        "    upper = len(chrom_df.index)\n",
        "    idx = int(upper/2)\n",
        "    chrom_idx = chrom_df.index\n",
        "\n",
        "    # Binary search\n",
        "    while True:\n",
        "      idx_old = idx\n",
        "      gene_start = int(chrom_df.loc[chrom_idx[idx],'start_pos'])\n",
        "      gene_end = int(chrom_df.loc[chrom_idx[idx],'end_pos'])\n",
        "\n",
        "      if mut_pos < gene_start:\n",
        "        # Mutation precedes gene, look in previous half\n",
        "        upper = idx\n",
        "        idx -= int((idx - lower) / 2)\n",
        "\n",
        "      elif mut_pos <= gene_end:\n",
        "        # Mutation within gene [start, end], store id\n",
        "        df.loc[i,'gene_id'] = chrom_df.loc[chrom_idx[idx],\"id\"]\n",
        "        df.loc[i,'gene_start'] = chrom_df.loc[chrom_idx[idx],\"start_pos\"]\n",
        "        df.loc[i,'gene_end'] = chrom_df.loc[chrom_idx[idx],\"end_pos\"]\n",
        "        df.loc[i,'strand'] = chrom_df.loc[chrom_idx[idx],\"strand\"]\n",
        "        df.loc[i,'frame'] = chrom_df.loc[chrom_idx[idx],\"frame\"]\n",
        "\n",
        "        hits += 1\n",
        "        break\n",
        "\n",
        "      else:\n",
        "        # Mutation follows gene, look in next half\n",
        "        lower = idx\n",
        "        idx += int((upper - idx) / 2)\n",
        "\n",
        "      if idx == idx_old: break # Mutation not within any gene\n",
        "\n",
        "  print(f\"Number of hits:\\n{hits}\\n\")\n",
        "  return df\n",
        "\n",
        "\n",
        "# Extract gene IDs and names with mutations\n",
        "def get_gene_dict(mut_id_df, gene_df):\n",
        "  gene_dict = {}\n",
        "  for gene in mut_id_df['gene_id']:\n",
        "    match = re.search(r'^ID=[A-Z0-9\\.]*',str(gene))\n",
        "    if match: \n",
        "      gene_dict[match[0]] = None\n",
        "\n",
        "  name_m = gene_df['id'].str.contains(\"Name=\")\n",
        "  name_df = gene_df[name_m]\n",
        "  for name in name_df['id']:\n",
        "    id_match = re.search(r'^ID=[A-Z0-9\\.]*',str(name))\n",
        "    name_match = re.search(r'(Name=)(.*)',str(name))\n",
        "    if id_match[0] in gene_dict.keys():\n",
        "      gene_dict[id_match[0]] = name_match[2]\n",
        "  return gene_dict\n",
        "\n",
        "\n",
        "# Apply all mutations in the mutation df to a specific gene id\n",
        "def insert_mutation(mutation_df, gene_id):\n",
        "\n",
        "  # Filter df by specific gene_id\n",
        "  mask = mutation_df['gene_id'].str.contains(gene_id)\n",
        "  df = mutation_df[mask]\n",
        "\n",
        "  # Define key variables\n",
        "  first_idx = df.index[0]\n",
        "  chrom = str(df.loc[first_idx,'CHROM']) + \"\\n\"\n",
        "  gene_start = int(df.loc[first_idx,'gene_start'])\n",
        "  gene_end = int(df.loc[first_idx,'gene_end'])\n",
        "  frame = int(df.loc[first_idx,'frame'])\n",
        "  chrom_start = genome_str.index(chrom) + len(chrom) - 1 # -1 because genome positions index starting with 1\n",
        "\n",
        "  # Find gene nt sequence in genome string\n",
        "  gene = genome_str[chrom_start + gene_start + frame: chrom_start + gene_end + 1]\n",
        "  old_seq = Seq(gene)\n",
        "\n",
        "  # Apply each mutation to gene nt sequence\n",
        "  for i in df.index:\n",
        "    mut_pos = int(df.loc[i,'POS1'])\n",
        "    mut_gene_pos = mut_pos - gene_start\n",
        "    \n",
        "    gene = gene[:mut_gene_pos] + str(df.loc[i,'ALT1']) + \\\n",
        "      gene[mut_gene_pos + len(str(df.loc[i,'REF1'])):]\n",
        "  \n",
        "  # Return new gene sequence (accounting for reverse strand)\n",
        "  new_seq = Seq(gene)\n",
        "  if str(df.loc[first_idx,'strand']) == '+':\n",
        "    return old_seq, new_seq, \"+\", frame\n",
        "  else:\n",
        "    return old_seq.reverse_complement(), new_seq.reverse_complement(), \"-\", frame"
      ],
      "metadata": {
        "id": "ZishkOCfGj0f",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Print names of genes with mutations\n",
        "\n",
        "mut_id_df = binary_mutation_search(mut_df_1, chromosome_dfs)\n",
        "\n",
        "# Remove non genes and print systematic gene names\n",
        "gene_dict = get_gene_dict(mut_id_df, genome_df)\n",
        "print(\"Systematic gene names:\\n\",np.array(list(gene_dict.keys())))\n",
        "print(\"\\nNumber of genes with mutations:\\n\",len(gene_dict.keys()))\n",
        "print(\"\\nKnown gene names (if in genome dataframe):\\n\",np.array(list(set(gene_dict.values()))))"
      ],
      "metadata": {
        "id": "sJ8kx05LOIC0",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Display mutation dataframe\n",
        "\n",
        "#Cut mutation dataframe to only mutations within genes\n",
        "mut_hit_df = mut_id_df.copy()\n",
        "mut_hit_df.drop('POS2',axis=1,inplace=True)\n",
        "mut_hit_df.drop('REF2',axis=1,inplace=True)\n",
        "mut_hit_df.drop('IN_FILE',axis=1,inplace=True)\n",
        "mut_hit_df.drop('ALT2',axis=1,inplace=True)\n",
        "for row in mut_hit_df.index:\n",
        "  if not \"ID\" in str(mut_hit_df.loc[row,\"gene_id\"]):\n",
        "    mut_hit_df.drop(row,inplace=True)\n",
        "\n",
        "if download_mutation_csv:\n",
        "  mut_hit_df.to_csv(job_name+\".csv\")\n",
        "mut_hit_df"
      ],
      "metadata": {
        "id": "aYTCWp1bPzq5",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Apply mutations to peptide sequences and write results to file\n",
        "\n",
        "def write_mutation_result(outfile, oldpep, newpep):\n",
        "  if oldpep == newpep:\n",
        "    outfile.write(\"Mutation did not cause an amino acid change\\n\")\n",
        "    outfile.write(f\"{newpep}\")\n",
        "  else:\n",
        "    outfile.write(f\"New peptide sequence:\\n{newpep}\\n\")\n",
        "  \n",
        "  # diff_count adapted from https://stackoverflow.com/questions/28423448/counting-differences-between-two-strings\n",
        "  diff_count = sum(1 for a, b in zip(str(newpep), str(oldpep)) if a != b) + abs(len(new_pep) - len(old_pep))\n",
        "\n",
        "  if diff_count < 20:\n",
        "    for i in range(np.min([len(oldpep),len(newpep)])):\n",
        "      if oldpep[i] != newpep[i]:\n",
        "        outfile.write(f\"AA#{i+1} {oldpep[i]} -> {newpep[i]}, \")\n",
        "    outfile.write(\"\\n\")\n",
        "    if len(oldpep) != len(newpep):\n",
        "      outfile.write(f\"Old and new peptide lengths do not match\\n\")\n",
        "  else:\n",
        "    for i in range(np.min([len(oldpep),len(newpep)])):\n",
        "      if oldpep[i] != newpep[i]:\n",
        "        first_diff = i+1\n",
        "        break\n",
        "    outfile.write(f\"\\nMore than 20 amino acids differ, mutation likely caused frameshift starting at AA#{first_diff}. Individual amino acid changes omitted for brevity.\\n\")\n",
        "\n",
        "\n",
        "outfile = open(job_name+\".txt\", \"w\")\n",
        "outfile.write(f\"Systematic gene names:\\n{np.array(list(gene_dict.keys()))}\\n\\n\")\n",
        "for gene in set(mut_hit_df['gene_id']):\n",
        "  old, new, strand, frame = insert_mutation(mut_hit_df,gene)\n",
        "  outfile.write(\"\\n--------------------------------------------------------------------------------\\n\\n\")\n",
        "  outfile.write(f\"Results for {gene} on {strand} in frame {frame}:\\n\\n\")\n",
        "\n",
        "  old_pep = old.transcribe().translate()\n",
        "  new_pep = new.transcribe().translate()\n",
        "  unshifted_stop_count = sum(1 for a in str(new_pep) if a=='*')\n",
        "\n",
        "  write_mutation_result(outfile, old_pep, new_pep)\n",
        "  \n",
        "  if not (\"exon:1\" in gene):\n",
        "\n",
        "    shift1 = len(old) % 3\n",
        "    shift2 = 3 - shift1\n",
        "    outfile.write(f\"\\nExon not at start of protein, recording results in other frames as well:\\n\")\n",
        "    \n",
        "    outfile.write(f\"\\nFrame {(frame+1)%3}:\\n\")\n",
        "    old_shift1 = old[1:]\n",
        "    new_shift1 = new[1:]\n",
        "    old_pep = old_shift1.transcribe().translate()\n",
        "    new_pep = new_shift1.transcribe().translate()\n",
        "    write_mutation_result(outfile, old_pep, new_pep)\n",
        "\n",
        "    outfile.write(f\"\\nFrame {(frame+2)%3}:\\n\")\n",
        "    old_shift2 = old[2:]\n",
        "    new_shift2 = new[2:]\n",
        "    old_pep = old_shift2.transcribe().translate()\n",
        "    new_pep = new_shift2.transcribe().translate()\n",
        "    write_mutation_result(outfile, old_pep, new_pep)\n",
        "    \n",
        "\n",
        "outfile.close()\n",
        "files.download(job_name+\".txt\")\n",
        "if download_mutation_csv:\n",
        "  files.download(job_name+\".csv\")"
      ],
      "metadata": {
        "id": "EhLKRfDb0Sgy",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}